src/main/java/com/google/cloud/spanner/r2dbc/SpannerResult.java

```
@@ -16,25 +16,84 @@
 
 package com.google.cloud.spanner.r2dbc;
 
+import com.google.cloud.spanner.ResultSet;
+import com.google.cloud.spanner.Struct;
+import com.google.cloud.spanner.r2dbc.util.Assert;
 import io.r2dbc.spi.Result;
 import io.r2dbc.spi.Row;
 import io.r2dbc.spi.RowMetadata;
+import java.util.Iterator;
 import java.util.function.BiFunction;
 import org.reactivestreams.Publisher;
 import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
 
 /**
  * {@link Result} implementation for Cloud Spanner.
+ *
  */
 public class SpannerResult implements Result {
 
+  private final ResultSet resultSet;
+
+  private final Mono<Integer> rowsUpdated;
+
+  /**
+   * Constructor for read-only query execution.
+   *
+   * @param resultSet the underlying result from Cloud Spanner.
+   */
+  public SpannerResult(ResultSet resultSet) {
+    this.resultSet = Assert.requireNonNull(resultSet, "A non-null ResultSet is required.");
+    this.rowsUpdated = Mono.just(0);
+  }
+
+  /**
+   * Constructor for DML query execution.
+   *
+   * @param rowsUpdated the number of rows affected by the operation.
+   */
+  public SpannerResult(int rowsUpdated) {
+    this.resultSet = null;
+    this.rowsUpdated = Mono.just(rowsUpdated);
+  }
+
   @Override
   public Publisher<Integer> getRowsUpdated() {
-    return null;
+    return this.rowsUpdated;
   }
 
   @Override
   public <T> Flux<T> map(BiFunction<Row, RowMetadata, ? extends T> f) {
-    return Flux.just(f.apply(null, null));
+
+    if (resultSet == null) {
+      return Flux.empty();
+    }
+
+    return Flux.fromIterable(() -> new Iterator<T>() {
+
+      private Boolean hasNext;
```

Can you just change this to:
`private boolean hasNext = resultSet.next();`
and then `hasNext()` is much simpler.

---

Also, can the map function be called more than once? Then we might have a problem.

---

good idea ! changed. 

Regarding the calling map more than once on a result: that's a good point. Perhaps we should throw an error if it's called twice?

---

What do other drivers like MySQL do? Do they have the ability to reset the cursor to the beginning?

---

Well, map() just returns a publisher. There is nothing wrong with acquiring a publisher multiple times. The interesting part is subscribing to the publisher from multiple places.

fromIterable [docs](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#fromIterable-java.lang.Iterable-) say it creates a new iterator for each subscriber; we might want to test what that means empirically for the spanner result set.

---

Ah, I see the problem; we call "next()" on iterator creation, so multiple publisher acquisitions are actually a problem.

---

I'd create a tracking issue for supporting multiple invocations of map(), and tackle it in a different PR.

---

I might be missing something here but I believe our hands are tied... The client lib's result set cannot be reset and we obviously do not want to cache the query results for a replay so I think that means we logically can only return an empty flux (what happens now) or throw an error on a second map attempt.

---

I think this is a good question to ask the R2DBC team for feedback.

---

There might be a way you can "replay" the stream or cache it:

Methods like flux.cache(); flux.replay() etc:
https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#cache--
https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#replay--

---

@dzou  yes of course we could cache pretty easily inside the SpannerResult too, but the point is that the result may be huge and we wouldn't' want to keep it around (which may be why the user wants to operate on the results one-at-a-time in the first place).

---

@mp911de Hey Mark, what's the policy or guideline on a `Result` being mapped over more than once?

---

Results are stream objects which can be consumed only once. More specifically: Rows are expected to be read from a stream source and so a `Row` can be consumed only once. Typically, there are buffers associated with a `Row` that get released after applying the mapping function.

Calling `Row.get(…)` is permitted multiple times to the same column.

---

Thank you Mark !